### PR TITLE
Adding filter functionality for recipes

### LINKED ISSUE #84: Adding filter function  recipes by category
Adding function to  filter recipes by categories  and function to  remove  the filter.

### PR DESCRIPTION
**Pull request by ohadrevach**
Adding function for users to  filter recipes by  categories.
Also added  function  to  remove the  filter and see  all the recipes

close #84

### CODE REVIEW

**Review by GalOved — DISMISSED** [review]
Nice work. 
This is looking very good.

File: tasties_app/views.py [hunk 1]
```diff
@@ -19,12 +19,39 @@ def base(request):
 
 def recipes(request):
     recipes_list = Recipe.objects.all()
+    categories_list = Category.objects.all()
+    selected_category = request.GET.get("category")
     recipes_with_ratings = {}
+
+    if selected_category != 'remove_filter' and selected_category:
```
Comment (GilShechter) [inline tasties_app/views.py:26]: In what case do we get `remove_filter`?

File: tasties_app/views.py [hunk 2]
```diff
@@ -19,12 +19,39 @@ def base(request):
...
+                        recipe_id=recipe
+                    ).aggregate(Avg("rating"))["rating__avg"]
+        recipes_with_ratings = OrderedDict(
+            sorted(recipes_with_ratings.items(), key=lambda x: x[1], reverse=True)
+        )
+        context = {
+            "recipes_with_ratings": recipes_with_ratings,
+            "categories_list": categories_list,
+        }
+        return render(request, "tasties_app/recipes.html", context)
```
Comment (GilShechter) [inline tasties_app/views.py:41]: You have a duplicate code here, I think you can make the code more readable if you remove lines 37-41 (they are the same as lines 50-54) and put lines 43-49 under `else` block.

**Review by GilShechter — CHANGES_REQUESTED** [review]
Nice work! I pulled the code and it looks great.
Added some comments below.
Did you think about adding tests?

**Comment by GilShechter** [comment]
Did you find any solution to put the categories in a slider or carousel?
Now it's just floating out of the page.

File: tasties_app/static/css/recipes.css [hunk 3]
```diff
@@ -51,3 +51,18 @@
     justify-content: space-between;
     padding: 10px 0 0 0;
 }
+#category_btn:hover{
```
Comment (GilShechter) [inline tasties_app/static/css/recipes.css:54]: I don't know why but all the CSS below doesn't work when I pull this PR.
Is it working for you?

File: tasties_app/templates/tasties_app/recipes.html [hunk 4]
```diff
@@ -5,6 +5,22 @@
...
+                        <button type="submit" class="btn" name="category" value="{{i}}" id="category_btn">{{i}}</button>
+                    </li>
+                {% endfor %}
+                </ul>
+            </form>
+        </div>
+    {% endblock categories_block %}
     <div class="grid">
     <div class="row">
     {% for recipe, rating in recipes_with_ratings.items %}
```
Comment (efraimvis) [inline tasties_app/templates/tasties_app/recipes.html:26]: Here we will completely remove `recipes_with_ratings.items`, in fact the line itself should be:
`{% for recipe, rating in recipes_list %}`

File: tasties_app/views.py [hunk 5]
```diff
@@ -19,12 +19,31 @@ def base(request):
 
 def recipes(request):
     recipes_list = Recipe.objects.all()
+    categories_list = Category.objects.all()
+    selected_category = request.GET.get("category")
     recipes_with_ratings = {}
```
Comment (efraimvis) [inline tasties_app/views.py:24]: We do not need this dictionary `recipes_with_ratings`. Instead we can query an annotated QuerySet, like this:
`recipes_list = Recipe.objects.annotate(recipe_rating=Avg('rating__rating')).order_by('-recipe_rating')` (of course, this will be the first definition of `recipes_list`)

File: tasties_app/views.py [hunk 6]
```diff
@@ -19,12 +19,31 @@ def base(request):
...
+    else:
+        for recipe in recipes_list:
+            recipes_with_ratings[recipe] = Rating.objects.filter(
+                recipe_id=recipe).aggregate(Avg("rating"))["rating__avg"]
+
+    recipes_with_ratings = OrderedDict(
+        sorted(recipes_with_ratings.items(), key=lambda x: x[1], reverse=True)
+    )
+    context = {
+        "recipes_with_ratings": recipes_with_ratings,
```
Comment (efraimvis) [inline tasties_app/views.py:43]: We should not return the dictionary, rather the annotated QuerySet `recipes_list`.

**Review by efraimvis — CHANGES_REQUESTED** [review]

**Review by efraimvis — DISMISSED** [review]

**Review by GalOved — DISMISSED** [review]
Good work. Looking good for me

**Review by GilShechter — APPROVED** [review]

File: tasties_app/views.py [hunk 7]
```diff
@@ -1,63 +1,84 @@
...
+            username = form.cleaned_data.get("username")
+            messages.success(request, "Account was created for " + username)
+            return redirect("login")
         else:
             messages.error(request, form.error_messages)
 
-    context = {'form': form}
-    return render(request, 'tasties_app/register.html', context)
+    context = {"form": form}
+    return render(request, "tasties_app/register.html", context)
```
Comment (liornoy) [inline tasties_app/views.py:84]: All of those changes i.e. switching `'` with `"` or the newlines in return render - make it harder to review this PR.
Please try to avoid it next time.

**Comment by liornoy** [comment]
Can you please add test for this new filtering feature? for example assert that when you make a Get request to 
`/?category=Breakfast` you get only the Breakfast recipes.

File: tasties_app/templates/tasties_app/recipes.html [hunk 8]
```diff
@@ -5,9 +5,25 @@
     <link rel="stylesheet" type="text/css" href="{% static 'css/recipes.css' %}" />
 {% endblock %}
 {% block content %}
+    {% block categories_block %}
+        <div class="container-fluid bg-light d-flex justify-content-center" id="category_btn_container">
+            <form method="get" action="{% url 'recipes' %}">
+                <button type="submit" class="btn" name="category" value="remove_filter" id="remove_filter">Remove Filter</button>
```
Comment (liornoy) [inline tasties_app/templates/tasties_app/recipes.html:11]: This will redirect to: `/?category=remove_filter`
I think we could make it cleaner by redirecting just to `/` and omitting the category query param altogether.

File: tasties_app/views.py [hunk 9]
```diff
@@ -1,63 +1,84 @@
...
-    recipes_with_ratings = OrderedDict(sorted(recipes_with_ratings.items(), key=lambda x: x[1], reverse=True))
-    context = {'recipes_with_ratings': recipes_with_ratings}
-    return render(request, 'tasties_app/recipes.html', context)
+    recipes_list = Recipe.objects.annotate(
+        recipe_rating=Avg("rating__rating")
+    ).order_by("-recipe_rating")
+    categories_list = Category.objects.all()
+    selected_category = request.GET.get("category")
+    if selected_category:
+        if selected_category != "remove_filter":
```
Comment (liornoy) [inline tasties_app/views.py:26]: Here we can just say - if there is no "category" query param - don't filter.
Without the "remove_filter" option.

File: tasties_app/views.py [hunk 10]
```diff
@@ -1,63 +1,84 @@
...
+            category = selected_category
+            try:
+                category = Category.objects.get(category_name=selected_category)
+            except Exception:
+                pass
```
Comment (liornoy) [inline tasties_app/views.py:31]: This will allow users to enter malicious URL addresses and we will still digest it and return a valid page.
I would rather that we will return a forbidden response here because the value of the category must be valid as it's hard-wired to the HTML button element.
And we should test this corner case too.

File: tasties_app/views.py (hunk 7 above)
Comment (efraimvis) [reply tasties_app/views.py:84]: @liornoy @ohadrevach This is probably pre-commit's handywork.

File: tasties_app/views.py (hunk 10 above)
Comment (efraimvis) [reply tasties_app/views.py:31]: @liornoy @ohadrevach Thanks for the information, didn't think this crossed anyone's mind.

File: tasties_app/views.py (hunk 7 above)
Comment (natifridman) [reply tasties_app/views.py:84]: Also, per PEP8, both ways work but you should be consistent

Comment (ohadrevach) [reply tasties_app/views.py:84]: It is the pre-commit's work, i'll give more attention for it next time.

File: tasties_app/views.py (hunk 9 above)
Comment (ohadrevach) [reply tasties_app/views.py:26]: we need to click on something in order to remove  the filters and displays all the recipes

**Review by efraimvis — CHANGES_REQUESTED** [review]
@ohadrevach Please rebase your local branch to match the state of the main repo.
This PR is conflicting with the `main` branch at the moment...

**Comment by natifridman** [comment]
@ohadrevach Any update on this?

**Review by GalOved — APPROVED** [review]

**Comment by ohadrevach** [comment]
> @ohadrevach Any update on this?

Added the test and rebased my branch

**Review by GalOved — APPROVED** [review]
LGTM

**Review by GilShechter — APPROVED** [review]
LGTM

**Review by efraimvis — APPROVED** [review]

**Review by svatares — APPROVED** [review]
